File: Python/CockTailSort.py
# Python program for implementation of Cocktail Sort


def cocktail_sort(a):
    n = len(a)
    swapped = True
    start = 0
    end = n - 1
    while swapped:

        # reset the swapped flag on entering the loop,
        # because it might be true from a previous
        # iteration.
        swapped = False

        # loop from left to right same as the bubble
        # sort
        for i in range(start, end):
            if a[i] > a[i + 1]:
                a[i], a[i + 1] = a[i + 1], a[i]
                swapped = True

        # if nothing moved, then array is sorted.
        if not swapped:
            break

        # otherwise, reset the swapped flag so that it
        # can be used in the next stage
        swapped = False

        # move the end point back by one, because
        # item at the end is in its rightful spot
        end -= 1

        # from right to left, doing the same
        # comparison as in the previous stage
        for i in range(end - 1, start - 1, -1):
            if a[i] > a[i + 1]:
                a[i], a[i + 1] = a[i + 1], a[i]
                swapped = True

        # increase the starting point, because
        # the last stage would have moved the next
        # smallest number to its rightful spot.
        start += 1


# Driver code to test above
a = [5, 1, 4, 2, 8, 0, 2]
cocktail_sort(a)
print('Sorted array is:', a)

File: JavaScript/CocktailSort.js
const cocktailSort = (arrInput) => {
	let startIndex = 0, endIndex = arrInput.length, isSwapped = true;
	while (isSwapped) {
		isSwapped = false;
		for (let i = startIndex; i < endIndex - 1; i++) {
			if (arrInput[i] > arrInput[i + 1]) {
				let temp = arrInput[i];
				arrInput[i] = arrInput[i + 1];
				arrInput[i + 1] = temp;
				isSwapped = true;
			}
		}
		endIndex--;
		if (!isSwapped)    break;
		isSwapped = false;
		for (let i = endIndex - 1; i > startIndex; i--) {
			if (arrInput[i - 1] > arrInput[i]) {
				let temp = arrInput[i];
				arrInput[i] = arrInput[i - 1];
				arrInput[i - 1] = temp;
				isSwapped = true;
			}
		}
		startIndex++;
	}
	return arrInput;
}
//Cocktail Sort is a variatio of bubble sort.
//Example Usage : 
// let myArr = [8, 6, -4, 1, 84, 35]
// console.log(cocktailSort(myArr))
